File: android/widget/src/main/kotlin/com/fatwabot/widget/OccasionWidget.kt
package com.fatwabot.widget

import android.content.Context
import androidx.compose.runtime.Composable
import androidx.compose.ui.unit.DpSize
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.glance.GlanceId
import androidx.glance.GlanceModifier
import androidx.glance.GlanceTheme
import androidx.glance.LocalSize
import androidx.glance.appwidget.GlanceAppWidget
import androidx.glance.appwidget.GlanceAppWidgetReceiver
import androidx.glance.appwidget.SizeMode
import androidx.glance.appwidget.provideContent
import androidx.glance.layout.Alignment
import androidx.glance.layout.Column
import androidx.glance.layout.Row
import androidx.glance.layout.Spacer
import androidx.glance.layout.fillMaxSize
import androidx.glance.layout.fillMaxWidth
import androidx.glance.layout.height
import androidx.glance.layout.size
import androidx.glance.layout.padding
import androidx.glance.text.FontWeight
import androidx.glance.text.Text
import androidx.glance.text.TextAlign
import androidx.glance.text.TextStyle
import com.fatwabot.core.common.DeepLink
import com.fatwabot.core.prayer.IslamicOccasion
import com.fatwabot.core.prayer.IslamicOccasionCalculator
import com.fatwabot.core.prayer.IslamicOccasionCountdown

private val WideCell = DpSize(250.dp, 110.dp)

/**
 * رمضان والعيد — days remaining to Ramadan and the two Eids. Mirror of iOS
 * `OccasionWidget`.
 *
 * Zero network and zero snapshot dependency: pure Umm al-Qura arithmetic, so it
 * renders correctly on a device that has never opened the app and never had a
 * location.
 */
class OccasionWidget : GlanceAppWidget() {
    override val sizeMode: SizeMode = SizeMode.Responsive(setOf(WideCell))

    override suspend fun provideGlance(context: Context, id: GlanceId) {
        val countdowns = IslamicOccasionCalculator.all()
        provideContent {
            GlanceTheme { OccasionContent(context, countdowns) }
        }
    }
}

@Composable
private fun OccasionContent(context: Context, countdowns: List<IslamicOccasionCountdown>) {
    // Narrow shows only the nearest occasion — three rows at that width is three
    // unreadable rows, and the nearest is the one being waited for.
    val compact = LocalSize.current.width < WideCell.width
    val shown = if (compact) countdowns.take(1) else countdowns

    Column(
        modifier = GlanceModifier
            .fillMaxSize()
            .brandSurface()
            .opensApp(context, DeepLink.PRAYER)
            .padding(12.dp),
        horizontalAlignment = if (compact) Alignment.CenterHorizontally else Alignment.Start,
        verticalAlignment = if (compact) Alignment.CenterVertically else Alignment.Top,
    ) {
        shown.forEach { item ->
            if (compact) {
                Text(
                    occasionLabel(context, item.occasion),
                    style = TextStyle(
                        color = MaroonProvider, fontSize = 14.sp, fontWeight = FontWeight.Bold,
                        textAlign = TextAlign.Center,
                    ),
                )
                Text(
                    "${item.daysRemaining}",
                    style = TextStyle(
                        color = MaroonProvider, fontSize = 34.sp, fontWeight = FontWeight.Bold,
                        textAlign = TextAlign.Center,
                    ),
                )
                Text(
                    context.getString(R.string.widget_occasion_days),
                    style = TextStyle(color = MutedProvider, fontSize = 11.sp),
                )
            } else {
                Row(modifier = GlanceModifier.fillMaxWidth()) {
                    Text(
                        occasionLabel(context, item.occasion),
                        style = TextStyle(color = InkProvider, fontSize = 13.sp),
                    )
                    Spacer(GlanceModifier.defaultWeight())
                    Text(
                        "${item.daysRemaining}",
                        style = TextStyle(
                            color = MaroonProvider, fontSize = 16.sp, fontWeight = FontWeight.Bold,
                        ),
                    )
                    Spacer(GlanceModifier.size(4.dp))
                    Text(
                        context.getString(R.string.widget_occasion_days),
                        style = TextStyle(color = MutedProvider, fontSize = 11.sp),
                    )
                }
                Spacer(GlanceModifier.height(8.dp))
            }
        }
    }
}

private fun occasionLabel(context: Context, occasion: IslamicOccasion): String = when (occasion) {
    IslamicOccasion.RAMADAN -> context.getString(R.string.widget_occasion_ramadan)
    IslamicOccasion.EID_AL_FITR -> context.getString(R.string.widget_occasion_eid_al_fitr)
    IslamicOccasion.EID_AL_ADHA -> context.getString(R.string.widget_occasion_eid_al_adha)
}

class OccasionWidgetReceiver : GlanceAppWidgetReceiver() {
    override val glanceAppWidget: GlanceAppWidget = OccasionWidget()
}

File: ios/Widgets/Sources/OccasionWidget.swift
import CoreKit
import PrayerKit
import SwiftUI
import WidgetKit

/// رمضان والعيد — days remaining to Ramadan and the two Eids.
///
/// Zero network and zero snapshot dependency: this is arithmetic over the
/// Umm al-Qura calendar, so it renders correctly on a device that has never
/// opened the app and never had a location.
struct OccasionEntry: TimelineEntry {
    let date: Date
    let countdowns: [IslamicOccasionCountdown]
}

struct OccasionProvider: TimelineProvider {
    /// Mirrors the Hijri adjustment the user set for the rest of the app, so the
    /// countdown cannot disagree with the Hijri date shown beside it.
    private var offsetDays: Int {
        UserDefaults(suiteName: "group.com.fatwabot.app")?
            .integer(forKey: "hijriOffsetDays") ?? 0
    }

    private func entry(at date: Date) -> OccasionEntry {
        OccasionEntry(
            date: date,
            countdowns: IslamicOccasionCalculator.all(from: date, offsetDays: offsetDays)
        )
    }

    func placeholder(in context: Context) -> OccasionEntry { entry(at: Date()) }

    func getSnapshot(in context: Context, completion: @escaping (OccasionEntry) -> Void) {
        completion(entry(at: Date()))
    }

    func getTimeline(in context: Context, completion: @escaping (Timeline<OccasionEntry>) -> Void) {
        let now = Date()
        // One entry per day for a week, so the number ticks down without the
        // system needing to wake the extension daily.
        let entries = (0..<7).compactMap { offset -> OccasionEntry? in
            Calendar.current.date(byAdding: .day, value: offset, to: now).map { entry(at: $0) }
        }
        let reload = Calendar.current.date(byAdding: .day, value: 6, to: now) ?? now
        completion(Timeline(entries: entries, policy: .after(reload)))
    }
}

struct OccasionWidget: Widget {
    var body: some WidgetConfiguration {
        StaticConfiguration(kind: "OccasionWidget", provider: OccasionProvider()) { entry in
            OccasionWidgetView(entry: entry)
                .brandWidgetContainer()
        }
        .configurationDisplayName(Text("widget.occasion.name"))
        .description(Text("widget.occasion.desc"))
        .supportedFamilies([.systemSmall, .systemMedium])
    }
}

struct OccasionWidgetView: View {
    @Environment(\.widgetFamily) private var family
    let entry: OccasionEntry

    var body: some View {
        // Small shows only the nearest occasion — three rows at that size is
        // three unreadable rows, and the nearest one is the one being waited for.
        let shown = family == .systemSmall
            ? Array(entry.countdowns.prefix(1))
            : entry.countdowns

        VStack(alignment: .leading, spacing: family == .systemSmall ? 2 : 6) {
            ForEach(shown, id: \.occasion.rawValue) { item in
                if family == .systemSmall {
                    Spacer(minLength: 0)
                    Text(LocalizedStringKey(item.occasion.titleKey))
                        .font(.headline)
                        .foregroundStyle(brandPrimary)
                    Text("\(item.daysRemaining)")
                        .font(.system(size: 40, weight: .bold, design: .rounded))
                        .monospacedDigit()
                        .foregroundStyle(brandPrimary)
                        .minimumScaleFactor(0.5)
                    Text("occasion.days")
                        .font(.caption2)
                        .foregroundStyle(brandMuted)
                    Spacer(minLength: 0)
                } else {
                    HStack {
                        Text(LocalizedStringKey(item.occasion.titleKey))
                            .font(.subheadline.weight(.medium))
                            .foregroundStyle(brandInk)
                        Spacer()
                        Text("\(item.daysRemaining)")
                            .font(.title3.weight(.bold))
                            .monospacedDigit()
                            .foregroundStyle(brandPrimary)
                        Text("occasion.days")
                            .font(.caption2)
                            .foregroundStyle(brandMuted)
                    }
                    if item.occasion.rawValue != shown.last?.occasion.rawValue {
                        Divider().overlay(brandMuted.opacity(0.25))
                    }
                }
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: family == .systemSmall ? .center : .top)
    }
}
